### PR TITLE
Harfanglab download file from endpoint action

### PR DESCRIPTION
**Pull request by Godefroy-Amaury**
Add the `action_harfanglab_download_file_from_endpoint` action.
Rely on the `downloadFile` job action.

### CODE REVIEW

**Review by jeromefellus-sekoia — COMMENTED** [review]
Looks great ! Thank you !

File: HarfangLab/harfanglab/job_executor.py [hunk 1]
```diff
@@ -1,20 +1,87 @@
...
-        return JobTriggerResult(**response.json()[0])
+
+        job_result = JobTriggerResult(**response.json()[0])
+
+        self._job_id = job_result.id
+        self._job_is_running = True
+
+        return job_result
+
+    def wait_for_job_completion(self) -> None:
```
Comment (jeromefellus-sekoia) [inline HarfangLab/harfanglab/job_executor.py:56]:
```suggestion
    def wait_for_job_completion(self) -> None: # pragma: no cover
```

**Review by jeromefellus-sekoia — CHANGES_REQUESTED** [review]

**Review by squioc — CHANGES_REQUESTED** [review]
Thanks for your contribution.

Some changes to apply:
- Can you increase the version number in the [manifest.json](https://github.com/SEKOIA-IO/automation-library/blob/main/HarfangLab/manifest.json) file?
- Can you add a new entry in the [CHANGELOG.md](https://github.com/SEKOIA-IO/automation-library/blob/main/HarfangLab/CHANGELOG.md) file?
Thank you.

File: HarfangLab/harfanglab/download_file_from_endpoint.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+        if not self._artefact_info_fetched:
+            raise RuntimeError("DownloadFileFromEndpointAction.fetch_artefact_info() not called")
+        if self._artefact_info is None:
+            # Note: This raise is mostly for semantic/mypy purpose and should never
+            # happen in real life (an error will always raise before)
+            raise RuntimeError(
+                "DownloadFileFromEndpointAction.fetch_artefact_info() has been called, "
+                "but '_artifact_info' is still None - Something really strange and wrong happen"
+            )
+        return self._artefact_info["sha256"]
```
Comment (squioc) [inline HarfangLab/harfanglab/download_file_from_endpoint.py:40]: When the file is not found, the `sha256` is undefined. Can you handle the case when the path to the file is incorrect or the download of the file failed?
Thanks

File: HarfangLab/harfanglab/download_file_from_endpoint.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+                f"(expected '{job_id}', got '{artefact_info['job_id']}')"
+            )
+
+        if artefact_info["agent"]["agentid"] != agent_id:
+            raise ValueError(
+                f"Given agent id and the one in the fetched artefact info missmatch "
+                f"(expected '{agent_id}', got '{artefact_info['agent']['agentid']}')"
+            )
+
+        self._artefact_info = artefact_info
```
Comment (squioc) [inline HarfangLab/harfanglab/download_file_from_endpoint.py:90]: I suggest that the `download_status` property in the result of the job must be verified and raise an error if some wnet wrong.

**Review by squioc — CHANGES_REQUESTED** [review]

File: HarfangLab/manifest.json [hunk 4]
```diff
@@ -26,7 +26,7 @@
   "name": "HarfangLab",
   "uuid": "8380240b-61a4-48b7-93e4-044a7ee2309b",
   "slug": "harfanglab",
-  "version": "1.23.1",
+  "version": "1.24",
```
Comment (squioc) [inline HarfangLab/manifest.json:29]: We follow the [semver](https://semver.org/) convention for version number.
```suggestion
  "version": "1.24.0",
```

File: HarfangLab/CHANGELOG.md [hunk 5]
```diff
@@ -7,6 +7,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## Unreleased
 
+- Added functionality to download a file on endpoint
```
Comment (squioc) [inline HarfangLab/CHANGELOG.md:10]: As you bump the version number, you can add it with the entry

```suggestion

## 2024-10-17 - 1.24.0

- Added functionality to download a file on endpoint
```

File: HarfangLab/harfanglab/download_file_from_endpoint.py (hunk 2 above)
Comment (squioc) [reply HarfangLab/harfanglab/download_file_from_endpoint.py:40]: Sample of the job response when the file is not found:

```json
{
  "count": 1,
  "next": null,
  "previous": null,
  "results": [
    {
      "shortname": "C:\\Users\\<REDACTED>\\Downloads\\hello.txt",
      "artefact_type": "downloaded file",
      "item_status": 0,
      "download_status": 1,
      "destination": "elastic",
      "job_id": "<REDACTED>",
      "@version": "1",
      "tenant": "<REDACTED>",
      "log_type": "investigation",
      "msg": "0 file(s) (0B) downloaded",
      "@timestamp": "2024-10-17T15:13:43.619423Z",
      "size": 0,
      "agent": {
        "ostype": "windows",
        "version": "4.1.6",
        "osversion": "10.0.22621",
        "osproducttype": "Windows 11 Pro",
        "domainname": "WORKGROUP",
        "agentid": "<REDACTED>",
        "hostname": "Windows11"
      },
      "date": "2024-10-17T15:13:42.975214Z",
      "job_instance_id": "<REDACTED>",
      "id": "<REDACTED>"
    }
  ],
  "filters": {
    "wildcard": [
      "artefact_type",
      "msg",
      "shortname",
      "longname",
      "agent.domainname",
      "agent.hostname",
      "agent.osproducttype"
    ]
  }
}
```

**Review by jeromefellus-sekoia — APPROVED** [review]

File: HarfangLab/CHANGELOG.md [hunk 6]
```diff
@@ -7,6 +7,11 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## Unreleased
 
+
+## 2024-10-17 - 1.24.0
+
+- Added functionality to download a file on endpoint
```
Comment (squioc) [inline HarfangLab/CHANGELOG.md:13]:
```suggestion
## 2024-10-17 - 1.24.0

### Added

- Added functionality to download a file on endpoint
```

**Review by squioc — APPROVED** [review]
LGTM. Just a little change to apply to the README.md

**Review by squioc — CHANGES_REQUESTED** [review]
A little change to apply in the README.md

**Review by squioc — APPROVED** [review]

**Pull request merged**